persistence/sql/persister_session.go

```diff
@@ -76,21 +90,50 @@ func (p *Persister) ListSessionsByIdentity(ctx context.Context, iID uuid.UUID, a
 	return s, nil
 }
 
+// UpsertSession creates a session if not found else updates.
+// This operation also inserts Session device records when a session is being created.
+// The update operation skips updating Session device records since only one record would need to be updated in this case.
 func (p *Persister) UpsertSession(ctx context.Context, s *session.Session) error {
 	ctx, span := p.r.Tracer(ctx).Tracer().Start(ctx, "persistence.sql.UpsertSession")
 	defer span.End()
 
 	s.NID = p.NetworkID(ctx)
 
-	if err := p.Connection(ctx).Find(new(session.Session), s.ID); errors.Is(err, sql.ErrNoRows) {
-		// This must not be eager or identities will be created / updated
-		return errors.WithStack(p.GetConnection(ctx).Create(s))
-	} else if err != nil {
-		return errors.WithStack(err)
-	}
+	return errors.WithStack(p.Transaction(ctx, func(ctx context.Context, tx *pop.Connection) error {
+		if err := sqlcon.HandleError(tx.Where("id = ? AND nid = ?", s.ID, s.NID).First(new(session.Session))); errors.Is(err, sqlcon.ErrNoRows) {
+			// This must not be eager or identities will be created / updated
+			if err := sqlcon.HandleError(tx.Create(s)); err != nil {
+				return err
+			}
+
+			for i := range s.Devices {
+				device := &(s.Devices[i])
+				device.SessionID = s.ID
+				device.NID = s.NID
+
+				if device.Location != nil {
+					device.Location = stringsx.GetPointer(stringsx.TruncateByteLen(*device.Location, SessionDeviceLocationMaxLength))
+				}
+				if device.UserAgent != nil {
+					device.UserAgent = stringsx.GetPointer(stringsx.TruncateByteLen(*device.UserAgent, SessionDeviceUserAgentMaxLength))
+				}
 
-	// This must not be eager or identities will be created / updated
-	return p.GetConnection(ctx).Update(s)
+				if err := sqlcon.HandleError(tx.Create(device)); err != nil {
+					return err
+				}
+
+				s.Devices[i] = *device
+			}
+
+			return nil
+		} else if err != nil {
```

Comment:
Doing this here will not modify the original error, so you return it in the next line without handling again :wink: 
You would have to do
```suggestion
		} else if err = sqlcon.HandleError(err);  err != nil {
```
which does not help with readability at all :sweat_smile: 
Also, wouldn't `exists` be false (i.e. zero value of bool) in case there is an error, and therefore you incorrectly create the session first? Moving error handling so far from where the error is returned is always tricky and error prone.

Reply:
Yep, you're right. I wasn't sure either, but was really thrown off by the original solution of checking if the error is `sqlcon.ErrNoRows`. I guess it comes down to experience in the codebase, because we do that a lot in general, but I had to really think about what that `if` statement does, which IMO is not a good sign...